documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,809 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+    VResize = 1 << 5,
+    Custom = 1 << 6,
+}
+```
+```cs
+[Flags]
+public enum CursorFlags
+{
+    None,
+    Confined = 1 << 0
 }
 ```
+
+# Keyboard Input
+
+Once again, the interface is very simple.
+
+```cs
+public interface IKeyboard : IInputDevice
+{
+    KeyboardState State { get; }
+    string? ClipboardText { get; set; }
+    void BeginInput();
+    void EndInput();
+}
+```
+
+`State` is the device state as defined earlier.
+
+`BeginInput` starts recording textual input, bringing up the on-screen keyboard on platforms where this is needed (i.e. mobile). Unlike 2.0, this is now required given that textual input is now on a full-string basis rather than a per-char basis, and we don't want to record textual input unless the user asks us to.
+
+For instance, instead of a `KeyChar` event being raised every time a character is pressed or repeated, a `TextInput` event is raised as the string is being built alleviating the need for developers to build the string themselves, given that the primary use of this event is for GUI input.
+
+`ClipboardText` allows getting and setting the text on the user's clipboard so they can paste information to/from your application in others.
+
+`KeyboardState` is defined as follows:
+
+```cs
+public readonly record struct KeyboardState
+(
+    string? Text,
+    KeyState Keys
+);
+```
+
+`Text` contains the text typed on the keyboard since `IKeyboard.BeginInput`. This is cleared (set to `null`) when `IKeyboard.EndInput` is called, and will not be non-`null` again until another `IKeyboard.BeginInput` call.
+
+```cs
+public readonly record struct KeyState
+(
+    InputReadOnlyList<Key> Down
+)
+{
+    public bool this[KeyName btn] { get; }
+    public bool this[int scancode] { get; }
+    public InputReadOnlyList<Key> Up { get; }
+}
+```
+
+The indexer returns `true` if a particular key is pressed, false otherwise. If the developer wishes to enumerate the key state, they must explicitly enumerate through even the `Down` buttons or `Up` buttons.
+
+`Up` and the indexer will be implemented in terms of `Down`, which is the only property that a backend will need to set. Scancode-only keys will not be present in `Up`.
+
+Note because not all keys are named, and because some developers may prefer to use scancodes instead, a `Key` struct is used instead of just having the list be a list of key names.
+
+```cs
+public readonly record struct Key(KeyName Name, int Scancode);
+```
+
+`KeyName` will be `Unknown` for scancode-only, unnamed keys.
+
+Changes to `KeyboardState` also have matching actor methods which are subject to the actor method rules i.e. the backend should call them in the order in which the backend received the events where possible etc (read the Actors section).
+
+```cs
+public interface IKeyboardActor : IActor
+{
+    void HandleKeyDown(ButtonEvent<IKeyboard, Key> @event);
+    void HandleKeyUp(ButtonEvent<IKeyboard, Key> @event);
+    void HandleTextInput(TextEvent<IKeyboard> @event);
```

Comment:
Wouldn't it make more sense to just leave `KeyChar` at this point? Even if we want to expose a `KeyboardState.Text` and have Silk build the string, it would make more sense and cause less implementation friction if each backend didn't have to copy-paste the same stringbuilder logic

Reply:
Agreed, will fix in a bit